### PR TITLE
Feature/31 create a group context

### LINKED ISSUE #31: Create a group context
## Abstract

create a group context

## Description

We defined a group context.
If you need the details, see the Nishiki project document.

## What should do

Implement a group object and the following:

- [x] #32 
- [x] #33

### PR DESCRIPTION
**Pull request by taniguchiiqqq**
## Overviews of implementation

- add Group aggregate root
- add GroupName value objects

## Review points
Do you think this code meets the group domain knowledge?

### CODE REVIEW

File: backend/main/test/Group/Domain/Entities/Group.test.ts [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+			it("attempt to add existing container", () => {
+				const changedGroup = group.addContainer(container);
+				expect(changedGroup.error).instanceOf(GroupDomainError);
+			});
+
+			it("remove Group container", () => {
+				const changedGroup = group.removeContainer(container.id).value!;
+				console.log("changedGroup", changedGroup.containers);
+
```
Comment (hitohata) [inline backend/main/test/Group/Domain/Entities/Group.test.ts]: Maybe, you forgot to delete this line.

File: backend/main/src/Group/Domain/Entities/Group.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
+import { Container, ContainerId } from "src/Group/Domain/Entities/Container";
+import { AggregateRoot, Err, Identifier, Ok, Result } from "src/Shared";
+import { DomainObjectError } from "src/Shared";
+import { User, UserId } from "src/User";
+
+interface IGroupProps {
+	name: string;
+	containers: Container[];
+	users: User[];
+}
```
Comment (hitohata) [inline backend/main/src/Group/Domain/Entities/Group.ts:11]: Are these enough of only IDs?

Comment (hitohata) [inline backend/main/src/Group/Domain/Entities/Group.ts:11]: Are these enough of only IDs?

**Review by hitohata — APPROVED** [review]
I'm not entirely sure, but it seems that the user and the container in the group only require IDs. Those are aggregates roots with their domain knowledge, which is not utilized within this group class.

This is just my humble opinion, but I believe the group should suffice with only the user ID and the container class ID, rather than including the entire class.

File: backend/main/test/Group/Domain/Entities/Group.test.ts (hunk 1 above)
Comment (taniguchiiqqq) [reply backend/main/test/Group/Domain/Entities/Group.test.ts]: oh noo, Thank you for finding this one 🙌

**Comment by taniguchiiqqq** [comment]
Thank you so much for your opinion! I think it has points and it's right. In order to be confident about modifying aggregates, I feel I should do research and learn how to use an aggregate root. Please give me some more time.

**Comment by taniguchiiqqq** [comment]
@hitohata 
I changed it following your advice. Please review it again.
I made an issue updating the domain model diagram in the document to catch up on these changes.
https://github.com/genesis-tech-tribe/nishiki-documents/issues/44

File: backend/main/src/Group/Domain/Entities/Group.ts [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+		return Group.create(this.id, {
+			...this.props,
+			name: name,
+		});
+	}
+
+	/**
+	 * Add containerId to the group.
+	 * If the containerId already exists in the group, return error.
+	 * @param container
```
Comment (hitohata) [inline backend/main/src/Group/Domain/Entities/Group.ts]: This comment should be "containerId".

File: backend/main/src/Group/Domain/Entities/Group.ts [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	 */
+	public addContainerId(
+		containerId: ContainerId,
+	): Result<Group, GroupDomainError> {
+		const matchedContainerId = this.props.containerIds.find(
+			(cid) => cid.equal(containerId) === true,
```
Comment (hitohata) [inline backend/main/src/Group/Domain/Entities/Group.ts]: This is preference but you don't need to check if it's true. The equal method returns a boolean value. You can write:
```typescript
~.find(cid => cid.equal(containerId))
```

File: backend/main/src/Group/Domain/Entities/Group.ts [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	/**
+	 * Remove containerId from the group.
+	 * If the containerId doesn't exist in the group, return error.
+	 * @param containerId
+	 */
+	public removeContainerId(
+		containerId: ContainerId,
+	): Result<Group, GroupDomainError> {
+		const containerIds = this.props.containerIds.filter(
+			(uid) => uid.equal(containerId) !== true,
```
Comment (hitohata) [inline backend/main/src/Group/Domain/Entities/Group.ts]: Same as the above.

File: backend/main/src/Group/Domain/Entities/Group.ts [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+		);
+		console.log("containerIds", containerId);
+		console.log("this.props", this.props.containerIds);
```
Comment (hitohata) [inline backend/main/src/Group/Domain/Entities/Group.ts]: You might forget to delete those two lines.

File: backend/main/src/Group/Domain/Entities/Group.ts [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+		return Group.create(this.id, {
+			...this.props,
+			containerIds: containerIds,
+		});
+	}
+
+	/**
+	 * Add userId to the group.
+	 * If the userId already exists in the group, return error.
+	 * @param user
```
Comment (hitohata) [inline backend/main/src/Group/Domain/Entities/Group.ts]: This is "userId".

File: backend/main/test/Group/Domain/ValueObjects/GroupName.test.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
+import { describe, expect, it } from "vitest";
+import { GroupName } from "../../../../src/Group/Domain/ValueObjects/GroupName";
```
Comment (hitohata) [inline backend/main/test/Group/Domain/ValueObjects/GroupName.test.ts]: You can change this import path. This is related to my mistake though...

**Review by hitohata — APPROVED** [review]
LGTM aside from comments!

**Comment by taniguchiiqqq** [comment]
@hitohata 
I really appreciate your careful review! I fixed them. I feel the power of reviewing to enhance the quality of the codebase.

**Pull request merged**